rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,775 @@
+        if isinstance(transformer_size, dict):
+            transformer_size = transformer_size[attribute]
+        if num_transformer_layers > 0:
+            self._tf_layers["transformer"] = TransformerEncoder(
+                num_layers=num_transformer_layers,
+                units=transformer_size,
+                num_heads=config[NUM_HEADS],
+                filter_units=transformer_size * 4,
+                reg_lambda=config[REGULARIZATION_CONSTANT],
+                dropout_rate=config[DROP_RATE],
+                attention_dropout_rate=config[DROP_RATE_ATTENTION],
+                sparsity=config[WEIGHT_SPARSITY],
+                unidirectional=config[UNIDIRECTIONAL_ENCODER],
+                use_key_relative_position=config[KEY_RELATIVE_ATTENTION],
+                use_value_relative_position=config[VALUE_RELATIVE_ATTENTION],
+                max_relative_position=config[MAX_RELATIVE_POSITION],
+                name=f"{attribute}_encoder",
+            )
+
+        self.output_units = self._calculate_output_units(
+            attribute, num_transformer_layers, transformer_size, config
+        )
+
+    def _calculate_output_units(
+        self,
+        attribute: Text,
+        num_transformer_layers: int,
+        transformer_size: int,
+        config: Dict[Text, Any],
+    ):
+        """Determine the output units based on which layer components are present.
+
+        The output units depend on which component is the last one in the internal 
+        pipeline that is `RasaInputLayer` -> `Ffnn` -> `Transformer`, because not all 
+        the components are necessarily created.
+        """
+        # transformer is the last component
+        if num_transformer_layers > 0:
+            return transformer_size
+        # the Ffnn block is the last component
+        elif len(config[HIDDEN_LAYERS_SIZES][attribute]) > 0:
+            # this is the output size of the last layer of the Ffnn block
+            return config[HIDDEN_LAYERS_SIZES][attribute][-1]
+        # only the RasaInputLayer is present
+        else:
+            return self._tf_layers["input_layer"].output_units
+
+    def _features_as_token_ids(
+        self, features: List[Union[tf.Tensor, tf.SparseTensor]]
+    ) -> Optional[tf.Tensor]:
+        """Creates dense labels (token IDs) used for negative sampling in MLM."""
+        # If there are dense features, we use them as labels - taking the first dense
+        # feature in the list because any dense feature will do the job.
+        for f in features:
+            if not isinstance(f, tf.SparseTensor):
+                return tf.stop_gradient(f)
+
+        # If no dense features are found, use a sparse feature but convert it into
+        # a dense one first.
+        for f in features:
+            if isinstance(f, tf.SparseTensor):
+                return tf.stop_gradient(self._tf_layers["sparse_to_dense_token_ids"](f))
+
+    def _create_mlm_variables(
```

Comment:
variable is not a good name here, since it has certain meaning in `tf` being a trainable variable.

Reply:
Well, then I'm changing `variables` to `tensors`. Those have certain meaning too but I can't stop using good names just because they also mean something in TF 😛